design-proposals/app-orch-deploy-applications.md

```diff
@@ -0,0 +1,208 @@
+and password. We could eliminate the need to have a separate `Registry` object by
+adding these three fields to the `Application` object:
+
+- `chart_url`. The URL of the OCI helm chart.
+
+- `chart_username`, `chart_password`. Optional credentials to use when fetching the Helm Chart.
+
+`chart_url` and the existing field `helm_registry_name` are mutually exclusive. The user must either
+choose the simplified approach, or choose the more flexible approach of using the `Registry` object.
+If `chart_url` is specified, then App Orch will also make it available as an image registry,
+consistent with how `Image Pull Secrets` are implemented in the deployment manager. As such,
+`chart_url` will also be mutually exclusive with `image_registry_name`.
+
+If the user has a more complex use case that requires a custom cert, or they want to abstract
+these details behind a named `Registry` object, they are free to do so. There is still value in
+the `Registry` object, particularly in the case where private registries with authentication
+are used as it allows sensitive security information to be stored in one object rather than
+being replicated in many. Folding registry into Application is an optional means of simplifying
+the common case, particularly for public charts.
+
+### Automated Helm Chart Import
+
+A separate proposal describes a `helm-to-dp` tool for automatically creating `Deployment Package`,
+`Application`, and `Registry` by fetching and inspecting a Helm Chart.
+This proposal works in synergy with that tool. The `helm-to-dp` tool currently creates all
+three objects. If this proposal is accepted, then the `helm-to-dp` tool
+will need to only create one object in the common case, instead of three.
+
+## Rationale
+
+One alternative approach would be eliminating `Deployment Package` entirely and make
+`Application` a hierarchical object (i.e. An application could contain child Applications).
+This approach has a certain elegance in that it mirrors the Helm pattern where a
+Helm Chart may contain child charts. However, this is seen as too much of a divergence from
+the current implementation as it would require completely refactoring the approach to
+packages. Nothing in this proposal is incompatible with this alternative -- it could be
+taken up as a second phase at a later time.
+
+We could also consider eliminating the `Deployment Package` entirely. If a user wants to
+deploy multiple charts at the same time, we could require them to use an "umbrella helm
+chart", and we would represent the umbrella chart as a single Application object. Eliminating
+objects is disruptive from a backward compatibility / upgrade standpoint. It also drops
+a capability that is currently in use with demonstrated value. We propose re-evaluating
+the prevelance of deployment packages after deployable Applications have been supported
+for a few release cycles, and then making a decision whether to deprecate them.
```

Comment:
I think this would greatly simplify both the product and the way people reason about the product.
Do we have a sense of how many of the DP we currently deal with contain more than one application? If most of the current mapping is 1:1 then maybe dropping DP support won't be as disruptive.

Reply:
The two cases that come to mind are Geti (around 9 applications) and Base-Extensions (12 applications). We would need to get PdM approval from Russel to remove it -- he and I were just brainstorming this week on how the AI Stacks could publish microservices that are distributed as "applications".